### PR TITLE
bugfix/8256 Multiple subscriptions part 2

### PR DESCRIPTION
**Pull request by connorwstein**
Long story around this:
1. We added https://github.com/smartcontractkit/chainlink/pull/3983 to prevent an error in a runlog subscription causing us to miss runlogs. 
2. Discovered the bug described here https://github.com/smartcontractkit/chainlink/pull/4289. That bug is a combination of 1. being present and also hitting a websocket read limit. 
3. That turned out to NOT solve the issue because when the head tracker disconnects, it also calls disconnect on the job subscriber which will empty the map of subscriptions. 

The fix presented here changes the model a little bit - instead of the head tracker driving the job subscriptions through its Connect/Disconnect calls, the Connect (made idempotent in 2.) only creates the subscriptions and since they auto-reconnect, we only shut them down when the job subscriber itself shuts down in Stop. A simpler fix would be to remove the reconnect logic added in 1. and just live with the risk that an error on the runlog subscription **without** a corresponding error in the head trackers subscription could stop the processing of runlog jobs. It seems plausible that such a scenario could happen though, which is why I think the current approach is safer.

### CODE REVIEW

File: core/services/eth/client.go [hunk 1]
```diff
@@ -24,6 +24,7 @@ import (
 )
 
 //go:generate mockery --name Client --output ../../internal/mocks/ --case=underscore
+//go:generate mockery --name Client --output mocks/ --case=underscore
```
Comment (connorwstein) [inline core/services/eth/client.go:27]: I believe we're generally trying to move to having mocks per package, starting that for the eth package here

File: core/services/subscription.go [hunk 2]
```diff
@@ -220,28 +221,30 @@ type ManagedSubscription struct {
...
 }
 
 // Unsubscribe closes channels and cleans up resources.
-func (sub ManagedSubscription) Unsubscribe() {
+func (sub *ManagedSubscription) Unsubscribe() {
 	if sub.ethSubscription != nil {
 		timedUnsubscribe(sub.ethSubscription)
 	}
-	close(sub.logs)
+	close(sub.done)
```
Comment (samsondav) [inline core/services/subscription.go:247]: Does sub.logs ever get closed?

File: core/services/subscription.go [hunk 3]
```diff
@@ -271,14 +274,15 @@ func (sub ManagedSubscription) listenToLogs(q ethereum.FilterQuery) {
 	backfilledSet := sub.backfillLogs(ctx, q)
 	for {
 		select {
-		case log, open := <-sub.logs:
-			if !open {
-				return
-			}
+		case <-sub.done:
+			return
```
Comment (samsondav) [inline core/services/subscription.go:278]: close sub.logs here?

File: core/services/subscription_internal_test.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+func (es EthSub) Unsubscribe() {
+	es.t.Log("unsubscribe called")
+}
+
+func (es EthSub) Err() <-chan error {
+	es.t.Log("err chan called")
+	return es.err
+}
+
+func TestSubscriptionListenToLogs(t *testing.T) {
```
Comment (samsondav) [inline core/services/subscription_internal_test.go:32]: I am more confident now I see a test. Did this fail on the previous code ?

**Review by samsondav — APPROVED** [review]

File: core/services/subscription_internal_test.go (hunk 4 above)
Comment (connorwstein) [reply core/services/subscription_internal_test.go:32]: Yes this failed on previous code - the fact that the we were updating a copy of sub.logs in the reconnect meant that the unsubscribe wouldn't work. However that is just sort of an unrelated landmine - the real fix is just to not empty the job subscription map in Disconnect here https://github.com/smartcontractkit/chainlink/pull/4342/files#diff-3aa9e1f0eaf11525321dfd8287eb6e975bfcc57899fdf2b79e2848ef48cd9b50L188

**Review by samsondav — APPROVED** [review]

**Pull request merged**